Quick Start for Existing Users V2 - Changes to existing task - Stats

AjeshRPai opened the pull request:
Fixes #16349 

This PR has the following changes 

#### Focus points on Quick Link Ribbon 
- For the below quick start tasks
   -  1. **Check your site stats**
   -  2. **Edit your Home page**
   -  3. **Review site pages**  

If the user is on dashboard tab, then the focus point is shown on the Quick Links ribbon

#### Stats 
- Updates the task title to `Check your site stats`
- Updates the task description to `Keep up to date with your site\'s performance`
- Updates Notice for Stats to `Select stats to see how your site is performing`

---

### To test:
Test 1 
- Launch the app with a site having a quick start in progress 
- Go to My Site -> Me -> App Settings -> Debug settings.
- Turn off the MySiteDashboardTabsFeatureConfig flag under the Features in development section.
Restart the app.
- Logout/login and select "Show me around" when the quick start prompt is shown.
- Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).
- Notice that for the following tasks - **Edit your home Page**, **Review Pages** and **Check stats**, the QuickStart focus point is shown in Site Menu Items 
- Notice that for the **Check stats** quick start task, the task title, description and Notice are updated as described in the PR

Test 2 & 3 - Prerequisites
- Go to My Site -> Me -> App Settings -> Debug settings.
- Enable `MySiteDashboardTabsFeatureConfig`
- Restart the app
   
Test 2 - When the default tab is `Site Menu` 
- Navigate to `My Site` -> `Me` -> `App Settings` 
- Tap `Initial Screen` and select `Menu` 
- Launch the app with a site having a quick start in progress 
- Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).
- Notice that for the following tasks - **Edit your home Page**, **Review Pages** and **Check stats**, the QuickStart focus point is shown in the Site menu list items. 
- Notice that for the quick start tasks  - **Check your Site, Social Sharing and Explore plans**, the Quickstart focus points are shown in the Site menu items as same as before. 
- Notice that for the **Check stats** quick start task, the task title, description and Notice are updated as described in the PR

Test 3 - When the default tab is `Home`
- Navigate to `My Site` -> `Me` -> `App Settings` 
- Tap `Initial Screen` and select `Home` 
- Launch the app with a site having a quick start in progress 
- Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).
- Notice that for the following tasks - **Edit your home Page**, **Review Pages** and **Check stats**, the QuickStart focus point is shown in the Quick links Ribbon Card. 
- Notice that for the quick start tasks  - **Check your Site, Social Sharing and Explore plans**, the Quickstart focus points is shown in the Site menu tab header first
- Notice that for the **Check stats** quick start task, the task title, description and Notice are updated as described in the PR
- On clicking the Site menu tab header, the list gets scrolled to show the Site menu item corresponding to the task 



---

### Known Issues
Adding the quick start focus point in the quick link ribbon reintroduces this bug #16288 


## Regression Notes
1. Potential unintended areas of impact
- Quick start focus points not shown properly 
- Stats Quick start task has incorrect title, description and notice 


2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests and Unit tests

3. What automated tests I added (or what prevented me from doing so)
Unit tests 

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

AjeshRPai committed:
Updates: Check your site stats title and subtitle

AjeshRPai committed:
Updates: resid and message to be final

AjeshRPai committed:
Revert "Remove: Focus point display logic from Quick Links ribbon"

This reverts commit 358602ce

AjeshRPai committed:
Removes: Site Menu step for stats, pages

AjeshRPai committed:
Adds: Field to enable and disable site menu item focus point

AjeshRPai committed:
Updates: String to show on prompt

AjeshRPai committed:
Fixes: Ktlint issues

AjeshRPai committed:
Updates: the name of the tests for given when then pattern

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt:
```diff
@@ -500,6 +504,14 @@ class MySiteViewModel @Inject constructor(
 
+    private fun enableQuickLinkRibbonFocusPoints(): Boolean {
+        return defaultABExperimentTab == MySiteTabType.DASHBOARD
+    }
```
1. Since we're not enabling the focus points here but just checking if they should be enabled, we can rename the function to reflect this.
2. Prefer using an expression body for functions with the body consisting of a single expression.

```suggestion
    private fun shouldEnableQuickLinkRibbonFocusPoints() = defaultABExperimentTab == MySiteTabType.DASHBOARD
```

P.S. Same applies to `enableSiteMenuFocusPoints()` function.

AjeshRPai replied:
@ashiagr Thanks for this suggestion. I completely agree with your suggestion. I have done the changes in the commit [ef5c2dd](https://github.com/wordpress-mobile/WordPress-Android/pull/16361/commits/ef5c2dd8ba00f91b11d496a01a4adff35d04a618).  Also I felt that the function name `shouldEnableSiteItemsFocusPoints` is better than `shouldEnableSiteMenuFocusPoints`. Wdyt?. 🤔

ashiagr replied:
I too like `shouldEnableSiteItemsFocusPoints()`. 🙂

ashiagr replied:
Verified in 2d22c65, ef5c2dd

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/items/SiteItemsBuilderTest.kt:
```diff
@@ -128,34 +128,54 @@ class SiteItemsBuilderTest {
 
     @Test
-    fun `passes parameter to show focus point to stats item`() {
+    fun `given stats focus point enabled, when card built, showFocusPoint should be true`() {
```
```suggestion
    fun `given stats focus point enabled, when card built, then showFocusPoint should be true`() {
```

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/items/SiteItemsBuilderTest.kt:
```diff
@@ -128,34 +128,54 @@ class SiteItemsBuilderTest {
+    }
+
+    @Test
+    fun `given stats focus point disabled, when card built, showFocusPoint should be false `() {
```
```suggestion
    fun `given stats focus point disabled, when card built, then showFocusPoint should be false`() {
```

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quicklinkribbons/QuickLinkRibbonBuilder.kt:
```diff
@@ -1,6 +1,7 @@
 package org.wordpress.android.ui.mysite.cards.quicklinkribbons
```
Nitpick: Not related to the changes but I noticed that pluralized name `quicklinkribbons` is used in the package name. It can be renamed to `quicklinksribbon`. Notice pluralized "links" and singular "ribbon".

Few other places in the codebase where pluralized `ribbons` still being used that we might want to rename:

`org.wordpress.android.ui.mysite.cards.CardsBuilder `-> `quickLinkRibbonsBuilderParams`

`org.wordpress.android.ui.mysite.cards.CardsBuilderTest` -> 
`quickLinkRibbons`
`quickLinkRibbonsBuilderParams`
`initQuickLinkRibbons`
`given tabs enabled, when cards are built, then quick link ribbons built`

AjeshRPai replied:
@ashiagr Thanks for looking into the package name and the occurrences of pluralized `ribbons`. Your observational skills are as always superb. 🙌🏼  I have done the necessary changes in 2f7963e. Please let me know if I need to make any additional changes. I believe we could update all the occurrences of`quickLinkRibbon` to `quickLinksRibbon`. But I think we can take it up as a separate PR.  Wdyt?. 🤔

ashiagr replied:
Sounds good! Feel free to merge this PR when you're ready.

ashiagr reviewed (approved):
Great job @AjeshRPai! I appreciate clear instructions and commits. Except for the known issue, I found everything working as expected. 🙌 

I'm approving this PR, just consider a few code-level nitpicks. 

Also since `release/19.7` code-freeze branch is created, I think it is safe to merge the PR after comments are addressed.

AjeshRPai committed:
Updates: with then expression.

Co-authored-by: ashiagr <agr.ash@gmail.com>

AjeshRPai committed:
Updates: the test expression with then

Co-authored-by: ashiagr <agr.ash@gmail.com>

AjeshRPai committed:
Updates: function expression in Viewmodel

Co-authored-by: ashiagr <agr.ash@gmail.com>

AjeshRPai committed:
Updates: function name and format

AjeshRPai committed:
Fixes: Ktlint and Detekt issues

AjeshRPai committed:
Updates: package name to quicklinksribbon

Renames: all occurrences of quick link ribbons to quick link ribbon.

AjeshRPai commented:
@ashiagr Thanks for the inputs and the review. Appreciate your suggestions. 👍🏼 🙌🏼

jostnes commented:
👋 @AjeshRPai, I tested the build, and I can see the text description for Stats task has been updated as described in the PR.

I did notice one thing for the Stats task though, if user is on the `Menu` tab and the QS task is on `Home` tab, there's no focus point asking the user to go to `Home` tab first:

https://user-images.githubusercontent.com/17252150/164372530-ef40e560-1d79-4e4e-a6ff-e444473bfd6e.mp4

Also do you how do we get the user to get QS on `Menu` instead of `Home`? So far, I've only been getting it on `Home` tab.

AjeshRPai commented:
hi @jostnes 👋🏼 

Thanks for testing the changes, 🙌🏼 .

>I did notice one thing for the Stats task though, if user is on the Menu tab and the QS task is on Home tab, there's no focus point asking the user to go to Home tab first:

Yup, you are right. This is a known [bug](https://github.com/wordpress-mobile/WordPress-Android/issues/16288) in the implementation. I have added it as a known issue in the PR description. Will take it up in a separate PR. 


>Also do you how do we get the user to get QS on Menu instead of Home? So far, I've only been getting it on Home tab.

You can change the default tab by following the steps 
- Navigate to `My Site` -> `Me` -> `App Settings` 
- Tap `Initial Screen` and select `Menu` 

Let me know if you need any help

jostnes commented:
> You can change the default tab by following the steps
> 
> 1. Navigate to My Site -> Me -> App Settings
> 2. Tap Initial Screen and select Menu

🙇 thanks for pointing this out! Completely missed that in the App Settings!

I tested the QS from Menu tab and found that the "auto-scroll" to focus point doesn't work on certain tasks, one of the tasks being `Stats`:

https://user-images.githubusercontent.com/17252150/164382984-5701e897-acf4-4d25-8596-3ce7346e6054.mp4

The auto-scroll is also not on the `Pages` menu item, for the `Edit your homepage` and `Review site pages` tasks. Let me know if this should be a separate issue or is something that you'd like to fix in this PR. Thanks!

AjeshRPai committed:
Updates: The logic of setting the selected tab

Note: This change will also update the my site dashboard screen after setting the initial screen. Previously the change was not reflected until the app restarted.

AjeshRPai committed:
Fixes: Format issues and updates comment

jostnes commented:
Retested the latest build and the auto-scroll issue on `Stats` and `Pages` related tasks works perfect now 🎉 :shipit:

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt:
```diff
@@ -1139,7 +1147,15 @@ class MySiteViewModel @Inject constructor(
 
     private fun selectDefaultTabIfNeeded() {
-        if (!isMySiteTabsEnabled || isDefaultABExperimentTabSet) return
+        if (!isMySiteTabsEnabled) return
+        // This logic checks if the current default tab is the same as the tab
+        // set as initial screen, if yes then return
+        if (isDefaultABExperimentTabSet) {
+            _selectTab.value?.let { tab ->
+                val currentDefaultTab = tab.peekContent().position
+                if (currentDefaultTab == orderedTabTypes.indexOf(defaultABExperimentTab)) return
+            }
+        }
         val index = orderedTabTypes.indexOf(defaultABExperimentTab)
         if (index != -1) {
```
Changes look good to me. Also I like that initial screen when changed from App Settings selects corresponding tab without restarting the app. 👏 One minor suggestion is to put common `defaultABExperimentTab` index logic at the top:  

```suggestion
        val index = orderedTabTypes.indexOf(defaultABExperimentTab)
        if (index != -1) {
            // This logic checks if the current default tab is the same as the tab
            // set as initial screen, if yes then return
            if (isDefaultABExperimentTabSet) {
                _selectTab.value?.let { tab ->
                    val currentDefaultTab = tab.peekContent().position
                    if (currentDefaultTab == index) return
                }
            }
```

Feel free to merge after this.

AjeshRPai replied:
@ashiagr  🤔  . I am sorry, I didn't understand actually the reason to shift the logic below. At present, the logic of checking whether the `isDefaultABExperimentTabSet`  is before accessing the index of the `defaultABExperimentTab `.  If we move the condition block with  `if (isDefaultABExperimentTabSet)` wouldn't the index of `defaultABExperimentTab` be accessed regardless of the value of `isDefaultABExperimentTabSet`.

AjeshRPai replied:
@ashiagr I have made the changes as per your suggestion in [56f442c](https://github.com/wordpress-mobile/WordPress-Android/pull/16361/commits/56f442c198739db9c0c4b4417fff084c6fd94dd6).

ashiagr replied:
LGTM, thanks, @AjeshRPai! 

:shipit:

AjeshRPai committed:
Updates: The logic of setting the selected tab

AjeshRPai committed:
Fixes: Formatting issues
